flake.nix

```diff
@@ -0,0 +1,65 @@
+{
+  description = "Higher-order, low-boilerplate free monads.";
+
+  inputs = {
+    nixpkgs.url = github:NixOs/nixpkgs/nixos-20.09;
+    flake-utils.url = github:numtide/flake-utils;
+  };
+
+  outputs = { nixpkgs, flake-utils, ... }:
+  flake-utils.lib.eachSystem ["x86_64-linux"] (system:
+  let
+    overrides = self: super:
+    let
+      c2n = name: src: self.callCabal2nixWithOptions name src "--no-hpack" {};
+      hack = pkg: ver: sha256: self.callHackageDirect { inherit pkg ver sha256; } {};
+    in {
+      polysemy = c2n "polysemy" ./.;
+      polysemy-plugin = c2n "polysemy-plugin" ./polysemy-plugin;
```

Comment:
I recommend replacing the raw paths you're using as source arguments (`./.` and `./polysemy-plugin`) with source that has had `.gitignore` taken into account. This will avoid unnecessary rebuilding.

e.g. replace `./.` with `pkgs.nix-gitignore.gitignoreSource [] ./.`.

You'll need to have Nixpkgs in scope (the `self` on line 25, not the `haskellPackages` `self` on line 12).

You can also add additional ignore patterns or ignore files to that list. In my projects, I like to have a `.nixignore` file used as a _whitelist_ instead of a blacklist. That way stuff like `README.md` or whatever doesn't change the hash for the Nix build.

Reply:
that's a pretty neat feature, I'll surely use that for my other projects!

I'm not sure what exactly needs to be in the build result – if you're only running a dev build, it's clear, but since you can export overlays with flake, for example, I'd assume that you'd want all relevant files to be in the package.

Also, wouldn't it be simpler to directly use `filterSource` if you're going to use an include list?

Reply:
Even if you're uncertain what is safe to exclude from the source, it would still be useful to clean out `.gitignore`-d files. But it's not a correctness issue omitting it, so it's up to you.